ray-curator/tests/backends/utils.py

```diff
@@ -0,0 +1,195 @@
+        df = input_data.to_pandas()
+        df["doc_length"] = df["text"].apply(len)
+        return DocumentBatch(
+            task_id=input_data.task_id,
+            dataset_name=input_data.dataset_name,
+            data=df,
+            _metadata=input_data._metadata,
+            _stage_perf=input_data._stage_perf,
+        )
+
+    @property
+    def name(self) -> str:
+        return "add_length"
+
+    def inputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], ["text"]
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], ["text", "doc_length"]
+
+
+class SplitIntoRowsStage(ProcessingStage[DocumentBatch, DocumentBatch]):
+    """Split the document into rows."""
+
+    @property
+    def resources(self) -> Resources:
+        return Resources(cpus=0.5)
+
+    def process(self, input_data: DocumentBatch) -> list[DocumentBatch]:
+        df = input_data.to_pandas()
+        # Remove source_files from metadata to prevent file collision issues
+        # When splitting a document into individual rows, each row would inherit
+        # the same source_files metadata. This causes the JsonlWriter to hash
+        # all rows to the same output file, resulting in overwrites instead of
+        # separate files per row. By removing source_files, we allow the writer
+        # to create unique output files for each row based on other metadata.
+        input_metadata_without_source_files = input_data._metadata.copy()
+        input_metadata_without_source_files.pop("source_files", None)
+
+        # Create a list of tasks, each with a unique task_id and metadata
+        tasks = []
+        for _, row in df.iterrows():
+            # Create a single-row DataFrame instead of a list of dicts
+            row_df = pd.DataFrame([row.to_dict()])
+            tasks.append(
+                DocumentBatch(
+                    task_id=f"{input_data.task_id}_row_{row['id']}",
+                    dataset_name=input_data.dataset_name,
+                    data=row_df,
+                    _metadata=input_metadata_without_source_files,
+                    _stage_perf=input_data._stage_perf.copy(),
+                )
+            )
+        return tasks
+
+    @property
+    def name(self) -> str:
+        return "split_into_rows"
+
+    @property
+    def ray_stage_spec(self) -> dict[str, bool]:
+        return {
+            "is_fanout_stage": True,
+        }
+
+    def inputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], []
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], []
+
+
+def create_test_pipeline(input_dir: Path, output_dir: Path) -> Pipeline:
+    """Create a test pipeline for integration testing."""
+    pipeline = Pipeline(
+        name="integration_test_pipeline", description="Integration test pipeline for backend comparison"
+    )
+
+    # Add JsonlReader stage
+    pipeline.add_stage(
+        JsonlReader(
```

Comment:
Would it be worth it to test `ParquetReader` here as well?

Reply:
This being an integration test for the backends, I don't see the need to test Jsonl & Parquet. Similarly for pandas and pyarrow. We can and should have separate tests for our `Readers` where we test that. But here the goal is an integration test that does a pipeline work with our backends, and the said pipeline should cover interesting situations such as fanout etc. Another "interesting" thing we can add later is one of the stages can be a gpu stage. But that I'll leave for a subsequent PR.